Clip peaks that are too far

lukipuki opened the pull request.

lukipuki committed:
Clip peaks that are too far

miskosz commented on python/covid_graphs/covid_graphs/predictions.py:
```diff
@@ -241,36 +243,50 @@ def load_prediction_db(prediction_dir: Path) -> PredictionDb:
+      'max_peak_variability'.
     """
     peak_days = [
         formula.get_peak(country_report=None) for formula in reversed(list(fitted_formulas))
```
I would prefer if we did not reverse the array, it makes the code more difficult to follow (you can iterate in reverse order by using `array[from:to:step]` notation with a negative step)

lukipuki replied:
I'm thinking that the API might not be the best right now, it was made for the previous use case which lived outside of `prediction.py`.

I suggest
- accepting `FittedFormula`'s in any order and we will sort them internally however we want.
- instead of an interval, we will return a subset of `fitted_formulas` with the desired properties.

This will simplify things and we would just construct the list instead of calculating the interval.

rejdi replied:
That will definitely help!

lukipuki replied:
I've now changed the API, PTAL.

I'm not inclined to simplify this further, since:
* this is really a TopCoder division 2 500 point problem, maybe not even that
* there are other parts of the code that need love right now :smile:

miskosz commented on python/covid_graphs/covid_graphs/predictions.py:
```diff
@@ -241,36 +243,50 @@ def load_prediction_db(prediction_dir: Path) -> PredictionDb:
+        skip += 1
+    clipped_peaks = peak_days[skip:]
+    if len(clipped_peaks) == 0:
+        return skip, skip
```
Nit: Consider `if skip == len(peak_days):` on the line before clipping.

lukipuki replied:
I've rewritten the code, so I'm not sure if this comment still applies.

miskosz reviewed (approved):
I have trouble wrapping my head around reversing, clipping and index counting, but I trust you that it works. (You can consider writing a test to earn bonus points.)

rejdi commented on python/covid_graphs/covid_graphs/predictions.py:
```diff
@@ -241,36 +243,50 @@ def load_prediction_db(prediction_dir: Path) -> PredictionDb:
-    for peak in peak_days[1:]:
+    for peak in clipped_peaks[1:]:
         min_peak = min(min_peak, peak)
         max_peak = max(max_peak, peak)
-        if max_peak - min_peak <= max_peak_distance:
+        if max_peak - min_peak <= max_peak_variability:
             take_count += 1
         else:
             break
-    return len(peak_days) - take_count
+    return len(clipped_peaks) - take_count, len(clipped_peaks)
```
Are clipped peaks monotone? If not, you may skip some predictions that are within interval!

lukipuki replied:
How do you mean your "you may"?

- "it's a bug, please don't do it" - I think this interpretation isn't likely, since we are returning a full interval without skipping predictions within the interval.
- "please create holes in the interval" - I think this would look weird on the slider.

rejdi replied:
Ok, now I understand what was the point.

lukipuki committed:
Improve the API of _calculate_displayable_predictions

Previously we returned indexes of an interval, now we're returning a
subsequence.

lukipuki commented:
Forgot to say that I've also added a test.
